util/ribbon_alg.h

```diff
@@ -803,18 +863,326 @@ bool SimpleFilterQuery(const typename FilterQueryHasher::Key &key,
 
 // InterleavedSolutionStorage is row-major at a high level, for good
 // locality, and column-major at a low level, for CPU efficiency
-// especially in filter querys or relatively small number of result bits
+// especially in filter queries or relatively small number of result bits
 // (== solution columns). The storage is a sequence of "blocks" where a
-// block has one CoeffRow for each solution column.
-
+// block has one CoeffRow-sized segment for each solution column. Each
```

Comment:
A query crossing the boundary between upper and lower number of columns gets rather ugly when more than two blocks might be involved. With just two blocks, once we get to memory loading, the whole query only sees one number of columns. We also have to worry about bit packing, e.g. 85 bits per column for 6 columns. I expect this to come with extra CPU overhead in extraction, though it might be OK if we're willing to generate custom code for each number of columns (up to ~16). Note also that come configurations like 9 columns have nontrivial unused data on the order of 1-2%.

Nevertheless, I should be able to test the hypothesis that number of cache lines matters by taking a nicely aligned setup like 4x128 and then deliberately misaligning that within cache lines so that each query accesses three cache lines instead of two. If number of cache lines really matters, we should see a difference in query times for skewed key distributions, where the number of keys' data fitting in various cache levels should matter.

I just hacked together such a test (based on prototype code connected to filter_bench) and it looks like we could attribute about a 10ns penalty per query for cache inefficiency. Based on some earlier prototypes and our simple storage, I expect extra CPU overhead to be at least that much, probably more like 20ns. So I don't think it's worth the extra code complexity.

Reply:
How about having just column-major solution structure. It would be good for negative queries, which I assume we care more about (do we have production metrics for that? I guess >90% bloom filter queries return negative?). The average memory access locations for a negative query is less than 2: (1/2 chance it only need to check 1 bit, 1/4 chance 2 bits, 1/8 chance 4 bits, ...).
The worst case would be bad, but as each rocksdb query checks several or even dozens sstables, it could average out.

Reply:
I prototyped that a while back and the non-locality made it expensive. (It might be reasonable with SIMD gather, e.g. _mm256_i64gather_epi64, but I don't want to depend on that for good performance.) The compiler and processor are in a "catch 22" situation where it either aggressively pre-fetches, potentially creating unnecessary TLB misses and cache loads, or has high memory latency to get next location. Sure, it has a similar cache warming conundrum for the sequential accesses in the interleaved layout, but the locality makes this case (IMHO) a completely ordinary cache warming conundrum.